File: docs-examples/example-java/src/test/java/io/micronaut/configuration/rabbitmq/docs/exchange/CustomExchangeSpec.java
package io.micronaut.configuration.rabbitmq.docs.exchange;

import io.micronaut.configuration.rabbitmq.AbstractRabbitMQTest;
import io.micronaut.context.ApplicationContext;
import org.junit.jupiter.api.Test;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

import static java.util.concurrent.TimeUnit.SECONDS;
import static org.awaitility.Awaitility.await;

public class CustomExchangeSpec extends AbstractRabbitMQTest {

    @Test
    void testUsingACustomExchange() {
        Map<String, Object> config = new HashMap<>();
        config.put("rabbitmq.port", rabbitContainer.getMappedPort(5672));
        config.put("spec.name", this.getClass().getSimpleName());
        ApplicationContext applicationContext = ApplicationContext.run(config, "test");

        AnimalClient client = applicationContext.getBean(AnimalClient.class);
        AnimalListener listener = applicationContext.getBean(AnimalListener.class);

        client.send("Cat", new Cat("Whiskers", 9));
        client.send("Cat", new Cat("Mr. Bigglesworth", 8));
        client.send("Snake", new Snake("Buttercup", false));
        client.send("Snake", new Snake("Monty the Python", true));
        Set<String> names = new HashSet<>();
        names.add("Whiskers");
        names.add("Mr. Bigglesworth");
        names.add("Buttercup");
        names.add("Monty the Python");

        await().atMost(10, SECONDS).until(() ->
                listener.receivedAnimals.size() == 4 &&
                listener.receivedAnimals.stream()
                        .filter(Cat.class::isInstance)
                        .anyMatch(cat -> cat.getName().equals("Whiskers") && ((Cat) cat).getLives() == 9) &&
                listener.receivedAnimals.stream()
                        .filter(Cat.class::isInstance)
                        .anyMatch(cat -> cat.getName().equals("Mr. Bigglesworth") && ((Cat) cat).getLives() == 8) &&
                listener.receivedAnimals.stream()
                        .filter(Snake.class::isInstance)
                        .anyMatch(snake -> snake.getName().equals("Buttercup") && !((Snake) snake).isVenomous()) &&
                listener.receivedAnimals.stream()
                        .filter(Snake.class::isInstance)
                        .anyMatch(snake -> snake.getName().equals("Monty the Python") && ((Snake) snake).isVenomous())
        );

        applicationContext.close();
    }
}

File: docs-examples/example-kotlin/src/test/kotlin/io/micronaut/rabbitmq/docs/exchange/CustomExchangeSpec.kt
package io.micronaut.rabbitmq.docs.exchange

import io.kotest.assertions.timing.eventually
import io.kotest.core.spec.style.AnnotationSpec
import io.kotest.matchers.collections.shouldExist
import io.kotest.matchers.shouldBe
import io.micronaut.context.annotation.Property
import io.micronaut.test.extensions.kotest5.annotation.MicronautTest
import io.micronaut.test.support.TestPropertyProvider
import jakarta.inject.Inject
import kotlin.time.Duration.Companion.seconds
import io.micronaut.rabbitmq.testcontainers.RabbitMQ

@MicronautTest
@Property(name = "spec.name", value = "CustomExchangeSpec")
class CustomExchangeSpec : TestPropertyProvider, AnnotationSpec() {

    override fun getProperties(): Map<String, String> {
        return RabbitMQ.getProperties()
    }

    @Inject
    lateinit var client: AnimalClient

    @Inject
    lateinit var listener: AnimalListener

    @Test
    suspend fun testUsingCustomExchange() {
        client.send(Cat("Whiskers", 9))
        client.send(Cat("Mr. Bigglesworth", 8))
        client.send(Snake("Buttercup", false))
        client.send(Snake("Monty the Python", true))

        val messages = listener.receivedAnimals
        eventually(10.seconds) {
            messages.size shouldBe  4
            messages shouldExist({ animal: Animal ->
                Cat::class.isInstance(animal) && animal.name == "Whiskers"
            })
            messages shouldExist({ animal: Animal ->
                Cat::class.isInstance(animal) && animal.name == "Mr. Bigglesworth"
            })
            messages shouldExist({ animal: Animal ->
                Snake::class.isInstance(animal) && animal.name == "Buttercup"
            })
            messages shouldExist({ animal: Animal ->
                Snake::class.isInstance(animal) && animal.name == "Monty the Python"
            })
        }
    }
}
